agent/src/util/device_plugin_service.rs

```diff
@@ -403,48 +455,62 @@ impl InstanceDevicePlugin {
-                Code::Unknown,
-                "Requested device already in use",
-            ))
+    instance_name: &str,
+    instance_namespace: &str,
+    capacity: &i32,
+    kube_interface: Arc<impl KubeInterface>,
+) -> Vec<(String, DeviceUsageStatus)> {
+    let mut device_usage_states = Vec::new();
+    match kube_interface
+        .find_instance(instance_name, instance_namespace)
+        .await
+    {
+        Ok(kube_akri_instance) => {
+            for (device_name, device_usage_string) in kube_akri_instance.spec.device_usage {
+                let device_usage_status = match DeviceUsage::from_str(&device_usage_string) {
+                    Ok(device_usage) => get_device_usage_state(&device_usage, node_name),
+                    Err(_) => DeviceUsageStatus::Unknown,
+                };
+                device_usage_states.push((device_name.clone(), device_usage_status));
+            }
+            device_usage_states
         }
-    } else {
-        // No corresponding id found
-        trace!(
-            "slot_available_to_reserve - could not find {} id in device_usage",
-            device_usage_id
-        );
-        Err(Status::new(
-            Code::Unknown,
-            "Could not find device usage slot",
-        ))
+        Err(_) => (0..*capacity)
+            .map(|x| {
+                (
+                    format!("{}-{}", instance_name, x),
+                    DeviceUsageStatus::Unknown,
+                )
+            })
+            .collect(),
     }
 }
 
+/// This returns device usage status of a `device_usage_id` slot for an instance on a given node
+/// # More details
+/// Cases based on the device usage value
+/// 1. DeviceUsageKind::Free ... this means that the device is available for use
+///     * (ACTION) return DeviceUsageStatus::Free
+/// 2. device_usage.usage == node_name ... this means node_name previously used device_usage
+///     * (ACTION) return previously reserved kind, DeviceUsageStatus::ReservedByConfiguration or DeviceUsageStatus::ReservedByInstance
+/// 3. device_usage.usage == (some other node) ... this means that we believe this device is in use by another node
+///     * (ACTION) return DeviceUsageStatus::ReservedByOtherNode
+fn get_device_usage_state(device_usage: &DeviceUsage, node_name: &str) -> DeviceUsageStatus {
+    let device_usage_state = match device_usage.get_kind() {
+        DeviceUsageKind::Free => DeviceUsageStatus::Free,
+        DeviceUsageKind::Configuration(vdev_id) => {
+            DeviceUsageStatus::ReservedByConfiguration(vdev_id)
+        }
+        DeviceUsageKind::Instance => DeviceUsageStatus::ReservedByInstance,
+    };
+    if device_usage_state != DeviceUsageStatus::Free && !device_usage.is_same_usage(node_name) {
+        return DeviceUsageStatus::ReservedByOtherNode;
+    }
+    device_usage_state
+}
+
 /// This tries up to `MAX_INSTANCE_UPDATE_TRIES` to update the requested slot of the Instance with the this node's name.
```

Comment:
update comment to more accurately describe what this function is doing.  for example, describe when errors are returned (if slot is owned by another node, slot is owned by another kind, slot is unknown) vs when/how the instance is updated.

Reply:
The doc comments describe what the function is doing.  I think it covers what you asked.

/// This returns device usage status of a `device_usage_id` slot for an instance on a given node
/// # More details
/// Cases based on the device usage value
/// 1. DeviceUsageKind::Free ... this means that the device is available for use
///     * (ACTION) return DeviceUsageStatus::Free
/// 2. device_usage.usage == node_name ... this means node_name previously used device_usage
///     * (ACTION) return previously reserved kind, DeviceUsageStatus::ReservedByConfiguration or DeviceUsageStatus::ReservedByInstance
/// 3. device_usage.usage == (some other node) ... this means that we believe this device is in use by another node
///     * (ACTION) return DeviceUsageStatus::ReservedByOtherNode

Reply:
sorry, meant comment for try_update_instance_device_usage

Reply:
add comments for try_update_instance_device_usage